File: Jump Game II - Leetcode 45/Jump Game II - Leetcode 45.js
// Brute Force
var jump = function(nums) {
    var n = nums.length;
    var smallest = [Infinity];

    function backtrack(i, jumps) {
        if (i === n - 1) {
            smallest[0] = Math.min(smallest[0], jumps);
            return;
        }

        var maxJump = nums[i];
        var maxReachableIndex = Math.min(i + maxJump, n - 1);

        for (var newIndex = maxReachableIndex; newIndex > i; newIndex--) {
            backtrack(newIndex, jumps + 1);
        }
    }

    backtrack(0, 0);
    return smallest[0];
};


// Optimal
var jump = function(nums) {
    var smallest = 0;
    var n = nums.length;
    var end = 0;
    var far = 0;

    for (var i = 0; i < n - 1; i++) {
        far = Math.max(far, i + nums[i]);
        
        if (i == end) {
            smallest++;
            end = far;
        }
    }

    return smallest;
};

File: Jump Game II - Leetcode 45/Jump Game II - Leetcode 45.java
// Brute Force
public class Solution {
    public int jump(int[] nums) {
        int n = nums.length;
        int[] smallest = {Integer.MAX_VALUE};

        backtrack(0, 0, nums, n, smallest);
        return smallest[0];
    }

    private void backtrack(int i, int jumps, int[] nums, int n, int[] smallest) {
        if (i == n - 1) {
            smallest[0] = Math.min(smallest[0], jumps);
            return;
        }

        int maxJump = nums[i];
        int maxReachableIndex = Math.min(i + maxJump, n - 1);

        for (int newIndex = maxReachableIndex; newIndex > i; newIndex--) {
            backtrack(newIndex, jumps + 1, nums, n, smallest);
        }
    }
}


// Optimal
class Solution {
    public int jump(int[] nums) {
        int smallest = 0;
        int n = nums.length;
        int end = 0;
        int far = 0;

        for (int i = 0; i < n - 1; i++) {
            far = Math.max(far, i + nums[i]);
            
            if (i == end) {
                smallest++;
                end = far;
            }
        }

        return smallest;
    }
}
